### ISSUE TITLE
native-image-agent omits String[] from "serialization" when using newConstructorForSerialization

### ISSUE THREAD

**Issue by Karm** [oracle/graal#9581]
As I was looking at https://github.com/oracle/graal/issues/8509#issue-2166454310 using native-agent to generate serialization config, it occurred to me that while e.g. both `java.lang.Short[]` and `java.lang.String[][]` get generated in the `reachability-metadata.json`, the `java.lang.String[]` is not added and has to be added manually. It's not a new thing though, I can see it both with older graalvm-community-jdk21u and with the very latest [graalvm-community-java24](https://github.com/graalvm/graalvm-ce-dev-builds/releases/download/24.2.0-dev-20240827_2122/graalvm-community-java24-linux-amd64-dev.tar.gz).

## How to reproduce
Using this source [Main.java.zip](https://github.com/user-attachments/files/16788770/Main.java.zip)
TL;DR:
```java
        Class[] arrayTypes = new Class[] {
...
                Short[][].class,
                String[].class,
                String[][].class,
...
        };
        for (int i = 0; i < arrayTypes.length; i++) {
            Constructor<?> cons = java.lang.Object.class.getDeclaredConstructor((Class[]) null);
            cons = ReflectionFactory.getReflectionFactory().newConstructorForSerialization(arrayTypes[i], cons);
            System.out.println(cons.getParameterCount());
            Object o = Array.newInstance(arrayTypes[i].getComponentType(), 0);
            System.out.println(o.getClass().descriptorString());
        }
```

```bash
$ native-image --version
native-image 24 2025-03-18
GraalVM Runtime Environment GraalVM CE 24-dev+12.1 (build 24+12-jvmci-b01)
Substrate VM GraalVM CE 24-dev+12.1 (build 24+12, serial gc)

$ java --version
openjdk 24 2025-03-18
OpenJDK Runtime Environment GraalVM CE 24-dev+12.1 (build 24+12-jvmci-b01)
OpenJDK 64-Bit Server VM GraalVM CE 24-dev+12.1 (build 24+12-jvmci-b01, mixed mode, sharing)

$ unzip Main.java.zip

$ javac Main.java

$ java -agentlib:native-image-agent=config-output-dir=AGENT Main

$ native-image -ea --no-fallback --link-at-build-time -H:ConfigurationFileDirectories=./AGENT Main

$ ./main
...
Exception in thread "main" com.oracle.svm.core.jdk.UnsupportedFeatureError: SerializationConstructorAccessor class not found for declaringClass: [Ljava.lang.String; (targetConstructorClass: java.lang.Object). Usually adding [Ljava.lang.String; to serialization-config.json fixes the problem.
    at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:121)
    at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.serialize.SerializationSupport.getSerializationConstructorAccessor(SerializationSupport.java:205)
    at java.base@24/jdk.internal.reflect.SerializationConstructorAccessorGenerator.generateSerializationConstructor(SerializationConstructorAccessorGenerator.java:53)
    at java.base@24/jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:340)
    at java.base@24/jdk.internal.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:253)
    at jdk.unsupported@24/sun.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:100)
    at Main.main(Main.java:83)
    at java.base@24/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
```

When I edit the `AGENT/reachability-metadata.json` file as suggested by the helpful message:

```diff
  {
    "type": "java.lang.Short[][]",
    "customTargetConstructorClass": "java.lang.Object"
  },
+ {
+   "type": "java.lang.String[]",
+   "customTargetConstructorClass": "java.lang.Object"
+ },
  {
    "type": "java.lang.String[][]",
    "customTargetConstructorClass": "java.lang.Object"
  },
```

and run the build again, it works.

Is this expected of the native agent?

(@zakkak FYI)

**Comment by fernando-valdez** [comment]
Thanks for reporting this issue. Created GR-57707 for internal tracking

**Comment by vjovanov** [comment]
Not expected behavior, I will look into this early next week.

**Comment by gingk1212** [comment]
Hi @vjovanov,

Looking into this issue, I identified the direct cause.
It seems that `[Ljava.lang.String;` is excluded due to the configuration in the following section, which prevents it from being recorded in reachability-metadata.json even when using the agent.

https://github.com/oracle/graal/blob/261c38a7a47dd87f13ec3cbe0fbb85cfdff17963/substratevm/src/com.oracle.svm.configure/src/com/oracle/svm/configure/trace/AccessAdvisor.java#L136-L137

The above line causes `shouldIgnore` method to return true, which results in it being skipped.

https://github.com/oracle/graal/blob/261c38a7a47dd87f13ec3cbe0fbb85cfdff17963/substratevm/src/com.oracle.svm.configure/src/com/oracle/svm/configure/trace/SerializationProcessor.java#L65-L67

As an experiment, I commented out the corresponding line in AccessAdvisor.java and confirmed that `[Ljava.lang.String;` was successfully added to reachability-metadata.json.

I hope this information is helpful.
